src/rgw/rgw_cache.h

```diff
@@ -26,6 +36,138 @@ enum {
 
 #define mydout(v) lsubdout(T::cct, rgw, v)
 
+/*engage1*/
+struct DataCache;
+class L2CacheThreadPool;
+class HttpL2Request;
+
+struct ChunkDataInfo : public LRUObject {
+	CephContext *cct;
+	uint64_t size;
+	time_t access_time;
+	string address;
+	string oid;
+	bool complete;
+	struct ChunkDataInfo *lru_prev;
+	struct ChunkDataInfo *lru_next;	
+
+	ChunkDataInfo(): size(0) {}
+
+	void set_ctx(CephContext *_cct) {
+		cct = _cct;
+	}
+
+	void dump(Formatter *f) const;
+	static void generate_test_instances(list<ChunkDataInfo*>& o);
+};
+
+struct cacheAioWriteRequest{
+	string oid;
+	void *data;
+	int fd;
+	struct aiocb *cb;
+	DataCache *priv_data;
+	CephContext *cct;
+	
+	cacheAioWriteRequest(CephContext *_cct) : cct(_cct) {}
+	int create_io(bufferlist& bl, unsigned int len, string oid);
+	
+	void release() {
+		::close(fd);
+		cb->aio_buf = NULL;
+		free(data);
+		data = NULL;
+		free(cb);
+		free(this);
+	}
+};
+
+struct DataCache {
+
+	private:
+		std::map<string, ChunkDataInfo*> cache_map;
```

Comment:
I skipped over this on my initial pass, but you might (or we might later) look at using boost::intrusive mapping for this--adds some complexity, but removes a lot of allocation;  examples of using it and its matching LRU in rgw_file.{h,cc}